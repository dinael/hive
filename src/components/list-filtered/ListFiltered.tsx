import {
  FC,
  ReactNode,
  useState,
  ChangeEvent,
} from 'react'

import { v4 as uuid } from 'uuid'

import styles from './ListFiltered.module.scss'

import ListFilters from './ListFilters'
import ListFilteredNotFound from './listFilteredNotFund'
import List from '../list/List'

export type ListFilteredItem = {
  id?: string | number
  text: string
}

export type ListFilteredProps = {
  items: ListFilteredItem[]
  nameList?: string
  className?: string
  children?: ReactNode
  OnlySearch?: boolean
  OnlySort?: boolean
  bullet?: 'none' | 'auto' | 'disc' | 'circle' | 'square' | 'decimal' | 'decimal-leading-zero' | 'lower-roman' | 'upper-roman' | 'lower-greek' | 'lower-latin' | 'upper-latin' | 'armenian' | 'georgian' | 'lower-alpha' | 'upper-alpha'
}

export const ListFiltered: FC<ListFilteredProps> = ({
  items,
  nameList,
  className = '',
  OnlySearch = false,
  OnlySort = false,
  bullet,
  children,
  ...props
}) => {
  const [searchQuery, setSearchQuery] = useState('');
  const [sortOption, setSortOption] = useState('top');

  if (!items) {
    return <div>Loading...</div>;
  }

  const filteredItems = items.filter(
    (item) =>
      item.text.toLowerCase().includes(searchQuery.toLowerCase())
  );

  const sortedItems = [...filteredItems].sort((a, b) => {
    if (sortOption === 'alphabetically_asc') {
      return a.text.localeCompare(b.text);
    } else if (sortOption === 'alphabetically_desc') {
      return b.text.localeCompare(a.text);
    }
    return 0;
  });

  const handleSortChange = (event: ChangeEvent<HTMLSelectElement>) => {
    setSortOption(event.target.value as typeof sortOption);
  };

  const optionsSelect = [
    {
      option: 'Default',
      value: 'default'
    },
    {
      option: 'A to Z',
      value: 'alphabetically_asc'
    },
    {
      option: 'Z to A',
      value: 'alphabetically_desc'
    }
  ];

  const showList = sortedItems.length > 0;

  return (
    <div
      className={`
        ${styles['list-filtered-component']}
        ${className || className}`}
        {...props}>
      <ListFilters
        className={`${styles['list-filter']}`}
        searchQuery={searchQuery}
        setSearchQuery={setSearchQuery}
        sortOption={sortOption}
        handleSortChange={handleSortChange}
        optionsSelect={optionsSelect}
        inputLabel='Search'
        selectLabel='Order by'
        title={nameList}
        onlySearch={OnlySearch}
        onlySort={OnlySort}
      />
      {children || (
        showList ? (
          <List
            items={sortedItems}
            aria-live='polite'
            ordered
            nameList={nameList ? nameList : `list-${uuid()}`}
            bullet={bullet ? bullet : 'none'}
          />
        ) : (
          <ListFilteredNotFound
            className={`${styles['list-not-found']}`}
            message='Search not found'
          />
        )
      )}
    </div>
  );
};

export default ListFiltered
