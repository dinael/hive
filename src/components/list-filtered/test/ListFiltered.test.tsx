import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import ListFiltered from '../ListFiltered';

describe('ListFiltered component', () => {
  const mockItems = [
    { id: 1, text: 'Item 1' },
    { id: 2, text: 'Item 2' },
    { id: 3, text: 'Item 3' },
  ];

  const renderComponent = (props = {}) => {
    render(<ListFiltered items={mockItems} {...props} />);
  };

  it('renders list items', () => {
    renderComponent();

    const item1 = screen.getByText('Item 1');
    const item2 = screen.getByText('Item 2');
    const item3 = screen.getByText('Item 3');

    expect(item1).toBeDefined();
    expect(item2).toBeDefined();
    expect(item3).toBeDefined();
  });

  // it('displays "Search not found" message when no items match the search', () => {
  //   renderComponent();

  //   const searchInput = screen.getByLabelText('Search') as HTMLInputElement;
  //   searchInput.value = 'Non-existing item';
  //   searchInput.dispatchEvent(new Event('input'));

  //   const notFoundMessage = screen.getByText('Search not found');
  //   expect(notFoundMessage).toBeDefined();
  // });

});
