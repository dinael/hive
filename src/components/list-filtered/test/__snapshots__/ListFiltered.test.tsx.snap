// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListFiltered > renders correctly 1`] = `
<div>
  <div
    class="
        list-filtered-component
        className-test"
  >
    <div
      class="list-filter"
    >
      <p
        class="list-filter-title"
      >
        List filtered
      </p>
      <label
        class="
        inputfiled-component
        
        list-filter-input
      "
      >
        <span
          class="
    label-text-component
    
    
    
    inputfiled-label
  "
        >
          Search
        </span>
        <input
          aria-describedby="validation-message-undefined"
          class=""
          placeholder="Search"
          spellcheck="false"
          type="search"
          value=""
        />
      </label>
      <label
        class="
    input-select-component
    
    list-filter-select
  "
        for=""
      >
        <span
          class="
    label-text-component
    
    
    
    input-select-label
  "
        >
          Order by
        </span>
        <select
          aria-describedby="validation-message-"
          id=""
        >
          <option
            value="default"
          >
            Default
          </option>
          <option
            value="alphabetically_asc"
          >
            A to Z
          </option>
          <option
            value="alphabetically_desc"
          >
            Z to A
          </option>
        </select>
        <span
          aria-hidden="true"
          class="
    icon-component
    xs
    icon
  "
        >
          <svg
            viewBox="0 0 24 24"
          >
            <title>
              chevron-down
            </title>
            <use
              xlink:href="/src/assets/sprite.svg#chevron-down"
            />
          </svg>
        </span>
      </label>
    </div>
    <ol
      aria-label="List filtered"
      aria-live="polite"
      class="
    list-component
    circle
    
    
  "
      role="list"
    >
      <li
        role="listitem"
      >
        Item 1
      </li>
      <li
        role="listitem"
      >
        Item 2
      </li>
      <li
        role="listitem"
      >
        Item 3
      </li>
    </ol>
  </div>
</div>
`;
